pkg/bpf/exporter.go

```diff
@@ -243,30 +232,36 @@ func (e *exporter) attach() error {
 		config.TaskClock:      {unix.PERF_TYPE_SOFTWARE, unix.PERF_COUNT_SW_TASK_CLOCK},
 	}
 
+	cleanup := func() error {
+		unixClosePerfEvents(e.perfEventFds)
+		e.perfEventFds = []int{}
+		e.enabledHardwareCounters = map[string]struct{}{}
+		return nil
+	}
+
 	for arrayName, counter := range counters {
 		bpfPerfArrayName := arrayName + bpfPerfArraySuffix
 		bpfMap, perfErr := e.module.GetMap(bpfPerfArrayName)
 		if perfErr != nil {
 			klog.Warningf("could not get ebpf map for perf event %s: %v\n", bpfPerfArrayName, perfErr)
-			continue
-		} else {
-			fds, perfErr := unixOpenPerfEvent(counter.EvType, counter.EvConfig, e.cpuCores)
-			if perfErr != nil {
-				// some hypervisors don't expose perf counters
-				klog.Warningf("could not attach perf event %s: %v. Are you using a VM?\n", bpfPerfArrayName, perfErr)
-				// if any counter is not enabled, we need disable HardwareCountersEnabled
-				e.hardwareCountersEnabled = false
-			}
-			for i, fd := range fds {
-				err = bpfMap.Update(unsafe.Pointer(&i), unsafe.Pointer(&fd))
-				if err != nil {
-					return fmt.Errorf("failed to update bpf map: %v", err)
-				}
+			return cleanup()
```

Comment:
earlier, we were skipping (continue) to the next counter, whereas here we are quitting the function. Is this intended?